alacritty/src/renderer/rectrenderer.rs

```diff
@@ -0,0 +1,258 @@
+
+            // VBO binding is not part ot VAO, but attributes are.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            // Position.
+            gl::VertexAttribPointer(
+                0,
+                2,
+                gl::FLOAT,
+                gl::FALSE,
+                (size_of::<Vertex>()) as _,
+                ptr::null(),
+            );
+            gl::EnableVertexAttribArray(0);
+
+            // Color.
+            gl::VertexAttribPointer(
+                1,
+                4,
+                gl::UNSIGNED_BYTE,
+                gl::TRUE,
+                (size_of::<Vertex>()) as _,
+                offset_of!(Vertex, color) as *const _,
+            );
+            gl::EnableVertexAttribArray(1);
+
+            // Reset buffer bindings.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self { vao, vbo, ebo, program, vertices: Vec::new() })
+    }
+
+    pub fn draw(&mut self, size_info: &SizeInfo, rects: Vec<RenderRect>) {
+        // Setup bindings. VAO will set up attribs and EBO, but not VBO.
+        unsafe {
+            gl::BindVertexArray(self.vao);
+
+            // Bind VBO only once for buffer data upload in draw_accumulated.
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(self.program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        for rect in &rects {
+            self.append_rect(center_x, center_y, rect);
+            if self.vertices.len() == MAX_VERTICES {
+                self.draw_accumulated();
+            }
+        }
+
+        if !self.vertices.is_empty() {
+            self.draw_accumulated();
+        }
+
+        unsafe {
+            // Disable program.
+            gl::UseProgram(0);
+
+            // Reset buffer bindings to nothing.
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+            gl::BindVertexArray(0);
+        }
+    }
+
+    fn append_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        // Make sure there's a space for at least 4 rectangle vertices.
+        debug_assert!(self.vertices.len() <= MAX_VERTICES - 4);
+
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
+        let color = Rgba {
+            r: rect.color.r,
+            g: rect.color.g,
+            b: rect.color.b,
+            a: (rect.alpha * 255.) as u8,
+        };
+
+        // Append rect vertices.
+        self.vertices.extend_from_slice(&[
+            Vertex { x, y, color },
+            Vertex { x, y: y - height, color },
+            Vertex { x: x + width, y, color },
+            Vertex { x: x + width, y: y - height, color },
+        ]);
+    }
+
+    fn draw_accumulated(&mut self) {
+        unsafe {
```

Comment:
That would also put vertices.clear() under unsafe, which i'm not sure i understand implications of

Reply:
In general, even if the entire code inside a function is safe it could still be unsafe, due to some external implications, so unsafe clearly exposes that to the user of the function, thus they'll be more carefully when using it. I'm not entirely sure if we should try to prevent with unsafe here, since the function is internal anyway...